### PR TITLE
Update functions

### PR DESCRIPTION
**Pull request by ikjudd**
- [x] refactored `wallet` and `connection ` params, pulling them off of the `client.provider` object
- [x] reassigned `program` var in each function to `await client.useProgram()`
- [x] updated logic in `hubs.js` to match logic of current working `nina-internal-sdk` and accommodate magic wallets
- [x] updated logic in `releaseCreate` and `releaseInitViaHub` in `release.js` to match logic of current working `nina-internal-sdk` and accommodate magic wallets
- [x] created functions for `exchange` - `exchangeInit()`, `exchangeAccept()` and `exchangeCancel()`
- [x] updated logic to invoke `program.methods` instead of `program.transaction` / `program.rpc`
- [x] updated docs to represent function implementation with examples
- [x] return objects from each function to the frontend

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@nina-protocol/js-sdk",
-  "version": "0.0.17",
+  "version": "0.0.14",
```
Comment (ejf89) [inline package.json]: should be updated to `0.0.18`

File: src/resources/exchanges.js [hunk 2]
```diff
@@ -1,41 +1,372 @@
...
+    tx.feePayer = provider.wallet.publicKey;
+    for await (let signer of [exchangeHistory]) {
+      tx.partialSign(signer);
+    }
+    const signedTx = await provider.wallet.signTransaction(tx);
+    const txid = await provider.connection.sendRawTransaction(signedTx.serialize(), {
+      skipPreflight: true,
+    });
+    await provider.connection.getParsedTransaction(txid, 'finalized');
+    return txid;
```
Comment (ejf89) [inline src/resources/exchanges.js]: should return object of accepted exchange

File: src/resources/exchanges.js [hunk 3]
```diff
@@ -1,41 +1,372 @@
...
+      .signers([])
+      .transaction();
+  }
+
+  tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash;
+  tx.feePayer = provider.wallet.publicKey;
+  const txid = await provider.wallet.sendTransaction(tx, provider.connection);
+  await provider.connection.getParsedTransaction(txid, 'confirmed');
+  return {
+    exchangePubkey,
```
Comment (ejf89) [inline src/resources/exchanges.js]: inconsistent return values

File: src/resources/hubs.js [hunk 4]
```diff
@@ -103,17 +123,805 @@ const fetchHubPost = async (publicKeyOrHandle, hubPostPublicKey, withAccountData
...
+      })
+      .transaction();
+    tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash;
+    tx.feePayer = provider.wallet.publicKey;
+    const txid = await provider.wallet.sendTransaction(tx, provider.connection);
+
+    await getConfirmTransaction(txid, provider.connection);
+    await axios.get(endpoints.api + `/hubs/${hubPublicKey}/collaborators/${hubCollaborator.toBase58()}`);
+    // endpoint needs to be updated to return collaborator pubkey
+    // const collaborator = await fetchHubCollaborator(hub.handle, collaboratorPubkey.toBase58());
```
Comment (ejf89) [inline src/resources/hubs.js]: should return `hubCollaborator`

File: src/resources/hubs.js [hunk 5]
```diff
@@ -103,17 +123,805 @@ const fetchHubPost = async (publicKeyOrHandle, hubPostPublicKey, withAccountData
...
+        collaborator: collaboratorPubkey,
+      })
+      .transaction();
+
+    tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash;
+    tx.feePayer = provider.wallet.publicKey;
+    const txid = await provider.wallet.sendTransaction(tx, provider.connection);
+
+    await getConfirmTransaction(txid, provider.connection);
+    // const collaborator = await fetchHubCollaborator(hub.handle, collaboratorPubkey.toBase58());
```
Comment (ejf89) [inline src/resources/hubs.js]: shoudl return `hubCollaborator`

File: src/resources/hubs.js [hunk 6]
```diff
@@ -103,17 +123,805 @@ const fetchHubPost = async (publicKeyOrHandle, hubPostPublicKey, withAccountData
...
+      })
+      .transaction();
+
+    tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash;
+    tx.feePayer = provider.wallet.publicKey;
+    const txid = await provider.wallet.sendTransaction(tx, provider.connection);
+
+    await getConfirmTransaction(txid, provider.connection);
+    await axios.get(endpoints.api + `/hubs/${hubPublicKey}/collaborators/${hubCollaborator.toBase58()}`);
+    // const collaborator = await fetchHubCollaborator(hub.handle, collaboratorPubkey.toBase58());
```
Comment (ejf89) [inline src/resources/hubs.js]: should return hubCollaborator

File: src/resources/releases.js [hunk 7]
```diff
@@ -76,160 +88,790 @@ const fetchExchanges = async (publicKey, withAccountData = false, pagination) =>
...
+      })
+      .preInstructions(instructions || [])
+      .transaction();
+
+    tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash;
+    tx.feePayer = provider.wallet.publicKey;
+    const txid = await provider.wallet.sendTransaction(tx, provider.connection);
+    await getConfirmTransaction(txid, provider.connection);
+
+    return txid;
```
Comment (ejf89) [inline src/resources/releases.js]: shoudl return release

File: src/resources/releases.js [hunk 8]
```diff
@@ -76,160 +88,790 @@ const fetchExchanges = async (publicKey, withAccountData = false, pagination) =>
...
+        releaseMint: release.releaseMint,
+        tokenProgram: TOKEN_PROGRAM_ID,
+      })
+      .preInstructions(instructions || [])
+      .transaction();
+    tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash;
+    tx.feePayer = provider.wallet.publicKey;
+    const txid = await provider.wallet.sendTransaction(tx, provider.connection);
+    await getConfirmTransaction(txid, provider.connection);
+    return txid;
```
Comment (ejf89) [inline src/resources/releases.js]: should return release

File: src/resources/releases.js [hunk 9]
```diff
@@ -76,160 +88,790 @@ const fetchExchanges = async (publicKey, withAccountData = false, pagination) =>
...
+        newRoyaltyRecipientTokenAccount,
+        tokenProgram: TOKEN_PROGRAM_ID,
+        rent: anchor.web3.SYSVAR_RENT_PUBKEY,
+      })
+      .transaction();
+    tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash;
+    tx.feePayer = provider.wallet.publicKey;
+    const txid = await provider.wallet.sendTransaction(tx, provider.connection);
+    await getConfirmTransaction(txid, provider.connection);
+    return recipientPublicKey;
```
Comment (ejf89) [inline src/resources/releases.js]: insonsistent return values

File: src/utils.js [hunk 10]
```diff
@@ -99,15 +105,15 @@ export const findAssociatedTokenAddress = async (
...
-
       let [usdcTokenAccountPubkey] = await findOrCreateAssociatedTokenAccount(
         connection,
         publicKey,
         publicKey,
         anchor.web3.SystemProgram.programId,
         anchor.web3.SYSVAR_RENT_PUBKEY,
-        new anchor.web3.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v')
-      )
+        // new anchor.web3.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v')
```
Comment (ejf89) [inline src/utils.js]: should not be here

**Review by ejf89 — CHANGES_REQUESTED** [review]
functions should return the resource that is relevant to the function.
JSdoc snippets should reflect that.

**Review by ejf89 — CHANGES_REQUESTED** [review]
functions should return the resource that is relevant to the function.
JSdoc snippets should reflect that.

**Comment by ikjudd** [comment]
this pass of updates to the repo focused on creating consistent return values. returning objects instead of strings
ex. `return { release: newReleaseData }`

**Comment by ikjudd** [comment]
merged `clean-up` branch into `update-functions`

File: src/client.js [hunk 11]
```diff
@@ -1,8 +1,48 @@
 import axios from 'axios';
 import * as anchor from '@project-serum/anchor';
 import _ from 'lodash';
+import Nina from '.';
```
Comment (facing-n) [inline src/client.js]: is this used anywhere?

File: src/resources/subscriptions.js [hunk 12]
```diff
@@ -34,7 +36,123 @@ const fetch = async (publicKey, withAccountData = false, transactionId = undefin
...
+
+tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash
+
+tx.feePayer = provider.wallet.publicKey
+
+const txid = await provider.wallet.sendTransaction(tx, provider.connection)
+await getConfirmTransaction(txid, provider.connection)
+const subscriptionData = await fetch(subscription.toBase58(), txid)
+return {
+  subscriptionData: subscriptionData,
```
Comment (facing-n) [inline src/resources/subscriptions.js]: return prop should be called name of resource it is returning - so here `subscription`

File: src/resources/subscriptions.js [hunk 13]
```diff
@@ -34,7 +36,123 @@ const fetch = async (publicKey, withAccountData = false, transactionId = undefin
...
+    .transaction()
+
+    tx.recentBlockhash = (await provider.connection.getRecentBlockhash()).blockhash
+    tx.feePayer = provider.wallet.publicKey
+    const txid = await provider.wallet.sendTransaction(tx, provider.connection)
+    await getConfirmTransaction(txid, provider.connection)
+
+    const subscriptionData = await fetch(subscription.toBase58(), txid)
+    return {
+      subscriptionData: subscriptionData,
```
Comment (facing-n) [inline src/resources/subscriptions.js]: `subscription`

**Review by facing-n — CHANGES_REQUESTED** [review]

**Review by ejf89 — APPROVED** [review]

**Pull request merged**